### PR TITLE
#19: Waiting for BucketFS content sync.

### LINKED ISSUE #19: Wait for BucketFS content sync
## Situation

After uploading a file to a bucket in BucketFS, synchronization mechanisms kick in. Users need to wait until sync is complete before using newly uploaded files.

## Acceptance criteria

* If the file is a archive of type `.tar.gz`, wait until the archive is unpacked.
* For other files wait until the archive until file appears in destination directory.

### PR DESCRIPTION
**Pull request by redcatbear**
Polls the BucketFS daemon log for messages that show that the object is synchronized.

Closes #19.

### CODE REVIEW

File: src/main/java/com/exasol/bucketfs/Bucket.java [hunk 1]
```diff
@@ -142,35 +176,69 @@ private String extractFirstPathComponent(final String path) {
...
     // [impl->dsn~uploading-to-bucket~1]
     public void uploadFile(final Path localPath, final String pathInBucket)
+            throws InterruptedException, BucketAccessException, TimeoutException {
+        uploadFile(localPath, pathInBucket, true);
+    }
+
+    /**
+     * Upload a file to the bucket and block the call until it is synchronized.
+     * <p>
+     * Uploads a file from a given local path to a URI pointing to a BucketFS bucket. If the bucket URI ends in a slash,
```
Comment (AnastasiiaSergienko) [inline src/main/java/com/exasol/bucketfs/Bucket.java]: I think this information should also be in the user guide

File: src/main/java/com/exasol/clusterlogs/LogPatternDetector.java [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    private static final int EXIT_OK = 0;
+    private static final Pattern LOG_ENTRY_PATTERN = Pattern
+            .compile("\\[. (\\d\\d)(\\d\\d)(\\d\\d) (\\d\\d:\\d\\d:\\d\\d).*");
+    private final Container<? extends Container<?>> container;
+    private final String logPath;
+    private final String pattern;
+    private final String logNamePattern;
+
+    /**
+     * Create a new instance of the {@link LogPatternDetector}
```
Comment (AnastasiiaSergienko) [inline src/main/java/com/exasol/clusterlogs/LogPatternDetector.java]:
```suggestion
     * Create a new instance of the {@link LogPatternDetector}.
```

File: src/main/java/com/exasol/containers/wait/strategy/LogFileEntryWaitStrategy.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+ * Strategy that waits for a container to be ready by checking when the language container is unpacked completely.
+ */
+public class LogFileEntryWaitStrategy extends AbstractWaitStrategy {
+    private static final long WAIT_DURATION_IN_MILLISECONDS = 60000;
+    private static final long POLLING_DELAY_IN_MILLISECONDS = 1000;
+    private final LogPatternDetector detector;
+    private final Instant afterUTC;
+
+    /**
+     * Create a new instance of the {@link LogFileEntryWaitStrategy}
```
Comment (AnastasiiaSergienko) [inline src/main/java/com/exasol/containers/wait/strategy/LogFileEntryWaitStrategy.java]:
```suggestion
     * Create a new instance of the {@link LogFileEntryWaitStrategy}.
```

File: src/main/java/com/exasol/containers/wait/strategy/LogFileEntryWaitStrategy.java [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    public LogFileEntryWaitStrategy(final LogPatternDetectorFactory detectorFactory, final String logPath,
+            final String logNamePattern, final String pattern) {
+        super();
+        this.afterUTC = Instant.now();
+        this.detector = detectorFactory.createLogPatternDetector(logPath, logNamePattern, pattern);
+    }
+
+    @Override
+    protected void waitUntilReady() {
+        final long expiry = System.currentTimeMillis() + (WAIT_DURATION_IN_MILLISECONDS);
```
Comment (AnastasiiaSergienko) [inline src/main/java/com/exasol/containers/wait/strategy/LogFileEntryWaitStrategy.java]: Why is WAIT_DURATION_IN_MILLISECONDS in brackets?

File: src/main/java/com/exasol/bucketfs/Bucket.java (hunk 1 above)
Comment (redcatbear) [reply src/main/java/com/exasol/bucketfs/Bucket.java]: Added section: Uploading Files Into a "Directory"

File: doc/user_guide/user_guide.md [hunk 5]
```diff
@@ -127,6 +127,18 @@ bucket.uploadStringContent(content, destination);
 
 Here `content` is the `String` that you want to write an destination is again the path inside the bucket.
 
+### Blocking vs. Non-blocking Upload
+
+In integration tests you usually want reproducible test cases. This is why the standard implementation of `uploadFile(...)` blocks the call until the underlying object is synchronized in the bucket.
+
+In rare cases you might want more control over that process, for example if you plan bulk-upload of a large number of small files and want to shift the check at the end of that operation.
```
[the remark this answers is not in the document]
Comment (redcatbear) [reply doc/user_guide/user_guide.md]: Fixed.

File: src/main/java/com/exasol/containers/wait/strategy/LogFileEntryWaitStrategy.java (hunk 4 above)
Comment (redcatbear) [reply src/main/java/com/exasol/containers/wait/strategy/LogFileEntryWaitStrategy.java]: Fixed.

**Pull request merged**